pkg/logql/ast.go

```diff
@@ -363,6 +364,41 @@ func (e *LineFilterExpr) Filter() (log.Filterer, error) {
 	}
 
 	return f, nil
+}*/
+
+func (e *LineFilterExpr) Filter() (log.Filterer, error) {
+
+	acc := make([]log.Filterer, 0)
+	for curr := e; curr != nil; curr = curr.Left {
+
+		switch curr.Op {
+		case OpFilterIP:
+			var err error
+			next, err := log.NewIPLineFilter(curr.Match, curr.Ty)
+			if err != nil {
+				return nil, err
+			}
+			acc = append(acc, next)
+		default:
+			next, err := log.NewFilter(curr.Match, curr.Ty)
+			if err != nil {
+				return nil, err
+			}
+			acc = append(acc, next)
+		}
+	}
+
+	if len(acc) == 1 {
+		return acc[0], nil
+	}
+
+	// The accumulation is right to left so it needs to be reversed.
+	for i := len(acc)/2 - 1; i >= 0; i-- {
+		opp := len(acc) - 1 - i
+		acc[i], acc[opp] = acc[opp], acc[i]
+	}
```

Comment:
I'm so happy one benchmark was slower. This would have been fatal. Anyhow, reversing the accumulation preserves the order. This is like depth first traversal with a stack except that the tree only goes down one side.